optuna/storages/_journal/storage.py

```diff
@@ -0,0 +1,612 @@
+import copy
+import datetime
+import enum
+import threading
+from typing import Any
+from typing import Container
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Sequence
+import uuid
+
+import optuna
+from optuna._experimental import experimental_class
+from optuna.distributions import BaseDistribution
+from optuna.distributions import check_distribution_compatibility
+from optuna.distributions import distribution_to_json
+from optuna.distributions import json_to_distribution
+from optuna.exceptions import DuplicatedStudyError
+from optuna.storages import BaseStorage
+from optuna.storages._base import DEFAULT_STUDY_NAME_PREFIX
+from optuna.storages._journal.base import BaseJournalLogStorage
+from optuna.study._frozen import FrozenStudy
+from optuna.study._study_direction import StudyDirection
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+_logger = optuna.logging.get_logger(__name__)
+
+NOT_FOUND_MSG = "Record does not exist."
+
+
+class JournalOperation(enum.IntEnum):
+    CREATE_STUDY = 0
+    DELETE_STUDY = 1
+    SET_STUDY_USER_ATTR = 2
+    SET_STUDY_SYSTEM_ATTR = 3
+    SET_STUDY_DIRECTIONS = 4
+    CREATE_TRIAL = 5
+    SET_TRIAL_PARAM = 6
+    SET_TRIAL_STATE_VALUES = 7
+    SET_TRIAL_INTERMEDIATE_VALUE = 8
+    SET_TRIAL_USER_ATTR = 9
+    SET_TRIAL_SYSTEM_ATTR = 10
+
+
+def datetime_from_isoformat(datetime_str: str) -> datetime.datetime:
+    # TODO(wattlebirdaz): Use datetime.fromisoformat after dropped Python 3.6 support.
+    return datetime.datetime.strptime(datetime_str, "%Y-%m-%dT%H:%M:%S.%f")
+
+
+@experimental_class("3.1.0")
+class JournalStorage(BaseStorage):
+    """Storage class for Journal storage backend.
+
+    Note that library users can instantiate this class, but the attributes
+    provided by this class are not supposed to be directly accessed by them.
+
+    Journal storage writes a record of every operation to the database as it is executed and
+    at the same time, keeps a latest snapshot of the database in-memory. If the database crashes
+    for any reason, the storage can re-establish the contents in memory by replaying the
+    operations stored from the beginning.
+
+    Journal storage has several benefits over the conventional value logging storages.
+    1. The number of IOs can be reduced because of larger granularity of logs.
+    2. Journal storage has simpler backend API than value logging storage.
+    3. Journal storage keeps a snapshot in-memory so no need to add more cache.
+
+    Example:
+
+        .. code::
+
+            storage = JournalStorage(JournalFileStorage("./log_file"))
+
+            study = optuna.create_study(storage=storage)
```

Comment:
[nits] How about adding `import` statements like the docstring of RedisStorage?
https://optuna.readthedocs.io/en/stable/reference/generated/optuna.storages.RedisStorage.html

Reply:
https://github.com/optuna/optuna/pull/3854/commits/6b5b0af6d137435ed8ab313dbc5bd568cd0318a8
https://github.com/optuna/optuna/pull/3854/commits/90a59b31766a0650f8055004e2a101c311111eeb